lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -1611,6 +1611,12 @@ LogicalResult InferenceMapping::mapOperation(Operation *op) {
         mappingFailed = true;
       });
 
+  // Forceable declarations should have the ref constrained to data result.
+  if (auto fop = dyn_cast<Forceable>(op); fop && fop.isForceable()) {
+    declareVars(fop.getDataRef(), fop.getLoc());
+    constrainTypes(fop.getDataRef(), fop.getDataRaw());
+  }
+
```

Comment:
This is interesting! I hadn't considered that the forceability would also influence width inference, but it definitely makes sense that it does. Does width inference also happen among the ref types? For example, if I have a wire with uninferred width, and I pass a ref to that up the module hierarchy and then force that ref somewhere up there, would the value I force it to influence the width inferred on the wire decl itself?

Reply:
Thanks for the consideration and great questions!  I'll try to answer, LMK if you have any more thoughts or it's not clear:

Widths only flow in the direction of the reference, never back the other way (even if forced) (this is in the FIRRTL spec, FWIW).  The intuition here is that the generated hardware should be the same with and without references, and "testing" code using them shouldn't (easily) influence the hardware even when not built separately (like GC taps).  IIRC it was important to allow taps to get their widths from their targets in the past.

So, here this should be saying "the rwprobe's width should be >= the data it's probing".  It's really "equal", but since references can only be "connected" to (`firrtl.ref.define`) once, statically, there's no worrying about inferring a larger width.

Reply:
That makes total sense! I love the restriction that the widths only flow from the decl to whatever is taking the reference, but you can't affect the generated hardware by forcing/referring to it. I should go read the spec 🙈